### ISSUE TITLE
Recovering Rook cluster after Kubernetes cluster loss

### ISSUE THREAD

**Issue by michalzxc** [rook/rook#1653]
Is this a bug report or feature request?

* Feature Request

**Feature Request**

Are there any similar features already existing:
n/a

What should the feature do:
Assuming I have backup of rook directory from nodes and I lost my kubernetes cluster I would like to be able to restore persistent volumes after kubernetes cluster redeployment making sure at the same time that existing PVs will be claimed by pods in new cluster

What would be solved through this feature:
Make possible to not lose all the data with lost of the kubernetes cluster

Does this have an impact on existing features:

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
Bare Metal/(VMs) with CoreOS
* Ceph status (use `ceph health` in the [Rook toolbox](https://Rook.io/docs/Rook/master/toolbox.html)):

**Comment by travisn** [comment]
This has come up several times, the question of how to recover the cluster when you lose the k8s cluster, but still have the ceph config and data persisted to disk. We'll start with docs around this, then eventually have tools to assist.

**Comment by pirogoeth** [comment]
Any updates on this?

**Comment by Queuecumber** [comment]
@travisn  It's extremely important for me to be able to bring a cluster up and down at a whim without losing our persistent which is currently in cephfs, so I'm wondering if you know of any method for doing this now or is it considered impossible?

**Comment by travisn** [comment]
I believe all the data needed could be found in the mon and osd config directories to restore the core cluster. What we will need is a tool that would extract the needed config and re-populate the ConfigMap and Secrets that Rook uses to keep track of the cluster. The critical resources to repopulate would include:
- rook-ceph-mon-endpoints (ConfigMap)
- rook-ceph-osd-<node>-config (ConfigMap)
- rook-ceph-mon (Secret)

Other artifacts likely can be regenerated by Rook, but it needs more investigation. Doing this manually would be a bit difficult, so would love to create a tool for this...

**Comment by Queuecumber** [comment]
If I get some time I might try to give this a shot, thought I'll need to set up an environment for testing it.. My initial thought is that I can use kubectl to dump the namespace that rook created to yaml, then re-create it on the new cluster. As long as the files in /var/lib/rook are still there and the devices are all correct it may just work. Then I would say bring rook itself up to take over the management. 

I suspect that for production youd want a more principled approach

**Comment by travisn** [comment]
@Queuecumber Great, it will be great to hear how the experiment goes!

**Comment by travisn** [comment]
One area to look at closely is the mon services. They should have the same ClusterIP or else it will change the identity of the mon and the mons will fail to form quorum. I'm hoping when you restore the K8s services in the `rook-ceph` namespace that it will allow them to keep their IPs.

**Comment by Queuecumber** [comment]
Thats interesting, I dont know much about the inner workings of ceph so I wouldnt have thought of that. If/when I give it a shot I will probablty set the cluser IP manually then, to whatever it gets after rook creates it the first time. For a production tool the previously assigned cluster IP would need to be stored in configuration somewhere so that it could automatically be applied when recreating the mon services

**Comment by bend** [comment]
I'm also looking for a tool to backup and restore my rook shared file system. This is a must have

**Comment by danielb42** [comment]
Any news on this? 

My scenario is that I often have to create new kubernetes clusters from scratch and want to use a shared cephfs from previous k8s clusters. 

The machines that I use as worker nodes have an always-persistent storage dir which I specified as dataDirHostPath in rook/ceph's cluster.yaml. Thus, those dirs contain `mon-x/, osdx/ and rook-ceph/` since I installed rook-ceph for the first time (by applying `operator.yaml, cluster.yaml and filesystem.yaml`). 

Now, trashing that whole k8s cluster and spinning up an all-new one, while only keeping the `dataDirHostPath`-location with its contents, the next k8s/rook/ceph-installation cannot mount the cephfs in `dataDirHostPath` as (so I guess) endpoint-IPs, configs/secrets and such are not matching any longer. I had no luck in updating any configs or secrets to the values that match the now-current cluster..

Any hints on how to proceed in such a scenario? Thanks.

**Comment by vistalba** [comment]
I think the scenario from daniel 42 is very importent. Also if one node has a disk failure and the dataDirHostPath is empty. (e.g. after disk replacement).

How to recover this situation?

**Comment by yan636** [comment]
This is a basic heal function that's critical for production.

**Comment by deeply** [comment]
Is it possible to read data directly from node disks , saying node machine and everything under `dataDirHostPath` are not corrupted.

**Comment by ndeixonne** [comment]
For some reasons, my rook cluster was deleted and I wanted to redeploy it.
Since /var/lib/rook (dataDirHostPath) contains data from the previous cluster deployment, it cannot be deployed.
I still have data on it and some PV are actually bounded to it. 

We will need a solution to recover a rook cluster also in this case.

Right now, we cannot store any sensitive data on Rook due to the recovery that is not clear.

**Comment by Queuecumber** [comment]
The more I think about this problem the more I really want to know why the data is stored in two places. From an engineering perspective it sounds like a really bad idea to have data in two places that needs to be kept in sync and which will result in corruption if either one changes without the other one being updated. 

Why not install the data to one place e.g. on disk and if rook detects that the configmaps are out of date, throw them out and rebuild them from what's on disk.

Ceph vanilla is extremely robust and dependable, but somehow using it through rook makes it extremely brittle. Let's say my cluster fails but i recover the disks. With rook in it's current state, even though I have all the disks and all the bytes are there, I have no recourse to recover the data. I don't know of another storage system that has this massive flaw.

**Comment by whereisaaron** [comment]
This conflict is because rook is not yet Kubernetes-native, it still partially relies on storage volumes and metadata that is external to the Kubernetes cluster and not managed by the cluster. Once rook starts using Kubernetes PV/PVC block storage instead of Linux block devices then Kubernetes will be able to reassemble Ceph clusters automatically, even after the complete deletion of the Kubernetes cluster (so long as the underly storage volumes still exist).

Right now a rook cluster can’t even survive Kubernetes cluster upgrade - which is a deletes and replaces all nodes in a rolling update. This is because it can’t automatically recover from even a single node failure/replacement. This is because Kubernetes doesn’t know about the storage volumes Ceph is secretly using, so it doesn’t know to remount them on the replacement nodes or to schedule the right OSD on that node. This realization was the end of our rook evaluation 😞 

In my view rook/Ceph needs to be entirely inside the Kubernetes (as a cluster service) or entirely outside the cluster (as an external storage service). At the moment rook is half-in, half-out. I think that is what makes it so brittle.

**Comment by laurybueno** [comment]
@whereisaaron I'm under the impression that Rook actually allows node replacement.

> Nodes can be added and removed over time by updating the Cluster CRD, for example with kubectl -n rook-ceph edit cephcluster rook-ceph. 
https://rook.io/docs/rook/v0.9/ceph-cluster-crd.html

Am I missing something?

**Comment by Queuecumber** [comment]
What happens when a node fails though? When I was evaluating if rook was production ready, I tried intentionally failing nodes. Sometimes it came back, sometimes I completely lost the storage cluster.

**Comment by noahdesu** [comment]
@whereisaaron there are currently in progress issues to back OSDs and monitors by PV/PVC. The rest of the daemon types do not require persistent storage, so this will be a big step forward.

@Queuecumber failing nodes will cause Ceph to take measures to prevent data loss such as making additional copies of data on other nodes. Of course this requires have spare capacity and redundancy in the cluster. If you were seeing a completely lost storage cluster, we'd certainly be interested in understanding those circumstances (in a new issue :)

**Comment by Queuecumber** [comment]
Consider the scenario that I was describing. I lose my hardware, lets say there's a fire in my datacenter or something worst-case scenario. Machines and SSDs (e.g. root drives) are hosed but my team pulls the disks and they're usable for recovery. I have all the bytes of my data that ceph was storing intact. Can I pop these drives in new machines and restore my customers data?

**Comment by noahdesu** [comment]
@Queuecumber ahh, that kind of recovery :) i'm not sure what the state of disaster recovery is from the pov of **rook**, so i think this thread should live on!

**Comment by Queuecumber** [comment]
Actually maybe not even disaster recovery, let's say I have my data stored in rook ceph, and for whatever reason I need to tear down and rebuild the cluster from scratch. E.g. I know exactly when and how the downtime will happen and I can plan accordingly and make backups of root drives/`dataDirHost`s / dump config maps whatever I need to do. Afaik there is not a way for me to do that and recover the data when the new cluster is up no matter how much I prepare.

**Comment by whereisaaron** [comment]
That's great news @noahdesu. I look forward to trying rook again when that hits a rook release. 🚀

@laurybueno true, and with the right process you can achieve an upgrade by incrementally adding a node, mount, zap, and configure the Linux device on the new node, delete an old node, wait for Ceph to reduplicate the deleted node data, then move to the next node. But then you end up copying the entire Ceph storage, including all the additional copies over everything.

With PV/PVC support is should be possible to 'drain' a node and have the OSD and PV automatically rescheduled to another node a few seconds later. Also, with the right right k8s health check and k8s [Pod Distruption Budget](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#how-disruption-budgets-work), you can ensure a rolling node update doesn't happen too fast, or automatically stops if there is a problem. Back to my robust, automated happy place while I 💤 and dream of ☕️ 😄 

@Queuecumber speaking from experience, you need to remove the circuit boards and have them dip-washed first to remove the smoke damage. And that is *after* sorting it out the assessment and claim with the insurance company. I'm saying it will still be faster to get offsite backups shipped to you 😃

**Comment by zerkms** [comment]
Please don't.

**Comment by kawaja** [comment]
If subscribers to this issue haven't noticed, @feisuzhu has improved the DR docs via PR https://github.com/rook/rook/pull/3555 which looks like it might satisfy this issue.

**Comment by Queuecumber** [comment]
Yeah kinda but it really should be a simpler process

**Comment by travisn** [comment]
@Queuecumber What suggestion do you have for a simpler process? Do you want something like a wiki where it's simple and quick to collaborate and share solutions? For the official documentation we like to review it thoroughly, but understood that it's heavyweight.

**Comment by Queuecumber** [comment]
What I'd like to see is migration of an existing rook-ceph cluster to a new kubernetes cluster supported as a first class feature not requiring a complex error prone procedure to pull off. I believe this is already the case for edgefs right?

**Comment by danielb42** [comment]
@travisn It does not satisfy the wider issue. While it's nice to know there is any procedure to recover a rook-ceph cluster (under certain conditions) this does not help so much when your platform spins up new (resp. resets) clusters frequently and automated. In such use cases, one cannot go through any manual steps to re-integrate stored data each time. So having a procedure of *26 manual steps* is a bit ridiculous.

I fail to see how rook, without such a feature, does not stand in contrast to the paradigms of container orchestration. It forces my infrastructure back into pet state. When the only solution currently available is to run a full-scale ceph, then what is the whole point of rook?

**Comment by Queuecumber** [comment]
Yeah that's very well said and exactly my issue as well. I'm actually not sure how rook can call itself 1.0 while missing this feature.

**Comment by Queuecumber** [comment]
@whereisaaron I'm curious if you tried edgefs with rook, I may be wrong but I think that can be migrated to a fresh cluster fairly hassle free

**Comment by whereisaaron** [comment]
No we didn't use EdgeFS except to poke at, but it still relies on the underlying structure of 'pet' nodes with [local `hostPaths` and Linux device names](https://rook.io/docs/rook/v1.1/edgefs-cluster-crd.html) so I think it has the same issues. I agree the multi-cluster support makes it easier to migrate data away from a cluster, but keeping it stable within each cluster still has the same challenges as far as I can see?

**Comment by Queuecumber** [comment]
Again I might be wrong but I was referring to the `devicesResurrectMode` setting of the Cluster CRD (https://rook.io/docs/rook/v1.1/edgefs-cluster-crd.html)

**Comment by whereisaaron** [comment]
@Queuecumber looks useful! I personally haven't tested if `devicesResurrectMode: restore` can recover a failed cluster, I'm more focused on it not failing so easily 😄

**Comment by travisn** [comment]
Ok, there are a number of different scenarios being discussed here. Hopefully this helps clarify what is available in Rook today, how v1.1 helps address some of these scenarios, what is still planned, and set expectations about what is really possible.

## 1. Disaster Recovery
If you unintentionally lose your K8s cluster, you should be able to recover the data. Absolutely, this is a critical feature of a reliable storage solution. Since Rook relies on K8s metadata, this currently requires some manual steps to recover as documented [here](https://rook.io/docs/rook/v1.1/disaster-recovery.html). Some thoughts on how to simplify/automate at least one scenario is captured in #3985. 

There is certainly much more to be done here. We need to automate this and reduce the manual steps as much as possible. It's an area of focus I'll be pushing for in the next release. But this won't ever be fully automated. Some manual steps will always be required.

## 2. Ephemeral Environments 
This category is for environments where it is expected that the K8s cluster is short-lived. There are a few variants of this...

### 2a. Nodes are added and removed frequently
Some K8s solutions have nodes that come and go. If you want to update a node, you just delete it and add a new one with the new version. Nodes are cattle and are expected to be completely stateless. 

The recommended approach in these clusters is to build the Rook cluster on top of PVs. With the v1.1 release, the Ceph mons and OSDs can be started on top of PVs. See the example [here](https://rook.io/docs/rook/v1.1/ceph-cluster-crd.html#pvc-based-cluster). As long as the PVs are "portable" and can be moved between nodes, the mons and OSDs can be rescheduled to new nodes when a node is drained. 

The portability of the PVs is very important. If you're running in a cloud environment, this is expected to work. If you're running on bare metal this obviously won't work. 

### 2b. Entire clusters removed and created again
Let's say there are entire clusters that are removed and expected to be restored again.  The challenge of course is that persistence and ephemeral really don't mix. In this environment, other solutions should really be explored than expecting a solution for full automation.  

Let's take AWS as an example. If you tear down the cluster and keep the EBS volumes around, how would Rook discover them again and be able to create the cluster on the same volumes? Rook doesn't have direct access to the EBS volumes. Rook relies on the EBS volume plugin from K8s to provision the volumes through the storage class. How can Rook re-map the same volumes? I'm all for exploring a solution for this if there is a clear need for it.

#### External Clusters
In v1.1 Rook supports separate clusters that can connect to each other for the storage. You can setup one cluster as the storage cluster, then the other clusters can connect to it for their storage needs. The other clusters can be ephemeral as needed, and the storage would be in a more stable cluster. Setting up the external cluster is described [here](https://rook.io/docs/rook/v1.1/ceph-cluster-crd.html#external-cluster).

#### Data Mirroring
There are several ways to mirror your data across clusters that essentially allows the data to be backed up off site.
- RBD Mirroring: See the `rbdMirroring` description in the [Cluster CR](https://rook.io/docs/rook/v1.1/ceph-cluster-crd.html#pvc-based-cluster). This does require extra steps to configure. Improvements are in progress for this. 
- RGW Multi-Site: Still a work in progress for Rook integration with #1584

**Comment by Queuecumber** [comment]
I disagree with your assessment of use case 2b, that is exactly how I want to use ceph. 

Speaking in general terms, our entire application stack is ephemeral and can be torn down to the point of rebuilding the root drives of the machines and rearchitected as necessary, but the underlying data that the application depends on shouldnt need to be rebuilt. We currently do this by having a dedicated storage machine as part of the cluster with a large zfs array and and nfs pod scheduled by kubernetes on it and it works fine but is a little inflexible. Ceph would make it so that the drives could be on any machine and we can get the same use case. As long as we dont touch non-root drives during the machine rebuilts, all the bytes of the data remain intact so this should be doable and my understanding is that with proper planning, using ceph external to rook we could do this. Rook would be ideal since we can schedule everything on kubernetes that way

**Comment by travisn** [comment]
@Queuecumber thanks for the feedback. I'm thinking of a way to make this scenario work. Would the following flow be possible for your scenario?
1. Before taking down the old cluster, export some k8s resources. For example, there are a few critical configmaps, secrets, and services that rook relies on for configuring the mons. 
1. After the new K8s cluster is started, import the resources before creating the rook cluster. If the host names change, some of these resources will need to be updated to map the old hosts to the new hosts.
1. Create the rook cluster. Rook will consume the resources previously imported instead of regenerating them. 
   - If the IP ranges changed, some intervention would be needed to restore mon quorum since the IP address is part of the mon identity.

The critical piece is really to get the basic mons up, using the same admin keyring and mon endpoints. The OSDs can then be discovered automatically in this case.  This would be interesting to prototype. Maybe it won't be this simple, but this would be ideal.

The disaster recovery steps assume that these resources in #1 aren't available, but if we have that metadata it becomes much more feasible.

**Comment by Queuecumber** [comment]
This pretty much exactly the flow I had in mind, our deployment manifests can even include the k8s resources that rook expects to have

**Comment by travisn** [comment]
@Queuecumber With a few minor fixes (see #4021), i was able to restore a cluster with the following steps. 
1. Start minikube (k8s 1.16) and attach two devices
1. Start a rook cluster. The default settings were used from common.yaml, operator.yaml, and cluster.yaml. The only change was that multiple mons per node were allowed since running in a single node (`allowMultiplePerNode: true`)
1. Wait for the cluster to be initialized and OSDs running. Create a pool and test consuming the storage.
1. Export the critical cluster settings
```bash
kubectl -n rook-ceph get secret rook-ceph-mon -o yaml  > rook-ceph-mon-secret.yaml
kubectl -n rook-ceph get secret rook-ceph-mons-keyring -o yaml  > rook-ceph-mon-keyring-secret.yaml
kubectl -n rook-ceph get secret rook-ceph-admin-keyring -o yaml > rook-ceph-admin-keyring-secret.yaml
kubectl -n rook-ceph get configmap rook-ceph-mon-endpoints -o yaml > rook-ceph-mon-endpoints-cm.yaml
kubectl -n rook-ceph get svc -l app=rook-ceph-mon -o yaml > rook-ceph-mon-services.yaml
```

5. Edit the exported yaml files to make them consumable in a new cluster by removing cluster-specific properties. For example:
   - Remove ownerReferences, resourceVersio, selfLink, uid
   - Add `namespace: rook-ceph` where missing
   - For the services, remove the list properties and add a separator `---` between each of the services (attached is a sample yaml after being modified)
6. Delete the Rook and/or K8s cluster (in my case I just deleted the Rook cluster.yaml, operator.yaml, and common.yaml to simulate a deleted cluster. It should have the same effect as bringing up a new k8s instance)
7. Create the new K8s cluster
8. Create the `rook-ceph` namespace and other basic resources: `kubectl create -f common.yaml`
9. Restore the Rook resources from the previous cluster
```
kubectl create -f rook-ceph-admin-keyring-secret.yaml
kubectl create -f rook-ceph-mon-keyring-secret.yaml
kubectl create -f rook-ceph-mon-services.yaml
kubectl create -f rook-ceph-mon-endpoints-cm.yaml
kubectl create -f rook-ceph-mon-secret.yaml
```
10. Create the Rook cluster: operator.yaml and cluster.yaml

If all goes well, the mons will restore quorum, the existing devices will be discovered and corresponding OSDs started, and the cluster will start successfully. The only issue I saw at this point was that some of the mgr commands were timing out so it took longer than expected. 

This should all be all capable of being automated. @Queuecumber want to give it a try after #4021 is merged?

**Comment by Queuecumber** [comment]
Very cool, I think this must have something to do with moving the metadata from host directories to k8s resources right? I'd love to do the implantation of this but probably someone with more experience with rook should go the design e.g. what kind of interface should be exposed to the user? Also how could we debug the commands which timeout?

**Comment by travisn** [comment]
This was prototyped based on the mons persisting data to the dataDirHostPath and the OSDs persisting to devices under /dev. If the scenario were to base these on PVCs, it would actually be more complicated since more k8s metadata would need to be restored and the PVs would need to be preserved.

I’ll continue investigating the mgr timeouts, just wanted to point it out in case someone tried this out. 

For improving the UX, what do you have in mind? Since the steps span k8s instances it would depend on a number of environmental factors.

**Comment by BlaineEXE** [comment]
It sounds to me like what we are talking about here is cluster adoption. Extending the ideas here to the most extreme case, all a user has is some number of OSD disks on some number of nodes from a prior Ceph cluster, and this cluster may not even have been a Rook cluster but a baremetal one. What is the minimum set of information needed to create a Ceph cluster that can use those disks as part of its storage? I think that would be the cluster FSID and the admin keyring. Or is it still required to have monitor data and the monitor keyring available? @leseb do you know?

But an intermediate step to this might be Rook having the ability to export and import data from Rook clusters, whether by a documented or automated process. IMO, a reasonably simple manual or script-based process for backing up and restoring a Rook cluster might make sense for the present, and we can set our minds to cluster adoption which would by necessity encompass all the features needed for all the various cases @travisn mentioned earlier. Ceph cluster adoption is non-trivial, and that will be a longer burn project.

**Comment by Queuecumber** [comment]
Yeah I like that idea, e.g. instead of having rook automatically set things up for me I could provide the ceph metadata to rook and it can take over a previously existing cluster

**Comment by travisn** [comment]
> It sounds to me like what we are talking about here is cluster adoption.

The question for cluster adoption is going to be around what is needed to get the mons to start and form quorum in rook. It might be possible already with the steps I outlined, but you would just have to create the yaml files yourself instead of first exporting them from a rook cluster. Once mon quorum is up, I expect Rook will pick up on the OSDs as long as they were created with `ceph-volume`. 

> But an intermediate step to this might be Rook having the ability to export and import data from Rook clusters, whether by a documented or automated process.

@BlaineEXE Are you suggesting exporting the full data in the rook cluster to another ceph cluster? RBD mirroring or rgw multi-site could help with this. But I think that is a separate topic from the question of how to restore a Rook cluster in a disaster recovery scenario. 

> IMO, a reasonably simple manual or script-based process for backing up and restoring a Rook cluster might make sense for the present, and we can set our minds to cluster adoption which would by necessity encompass all the features needed for all the various cases @travisn mentioned earlier. Ceph cluster adoption is non-trivial, and that will be a longer burn project.

Right, this is what I'm suggesting... A simple solution for backing up rook metadata and then being able to restore it. Keeping the cluster adoption as a separate topic sounds good.

**Comment by albertdb** [comment]
@travisn Would that prototype recover PVCs/PVs consuming Rook Ceph storage? The aim I think should be allowing using certain PVC (`mypersistentvolume`) from a Pod in the new Kubernetes cluster just as if it was deployed in the old cluster.

**Comment by umagmrit** [comment]
> @Queuecumber With a few minor fixes (see #4021), i was able to restore a cluster with the following steps.
> 
> 1. Start minikube (k8s 1.16) and attach two devices
> 2. Start a rook cluster. The default settings were used from common.yaml, operator.yaml, and cluster.yaml. The only change was that multiple mons per node were allowed since running in a single node (`allowMultiplePerNode: true`)
> 3. Wait for the cluster to be initialized and OSDs running. Create a pool and test consuming the storage.
> 4. Export the critical cluster settings
> 
> ```shell
> kubectl -n rook-ceph get secret rook-ceph-mon -o yaml  > rook-ceph-mon-secret.yaml
> kubectl -n rook-ceph get secret rook-ceph-mons-keyring -o yaml  > rook-ceph-mon-keyring-secret.yaml
> kubectl -n rook-ceph get secret rook-ceph-admin-keyring -o yaml > rook-ceph-admin-keyring-secret.yaml
> kubectl -n rook-ceph get configmap rook-ceph-mon-endpoints -o yaml > rook-ceph-mon-endpoints-cm.yaml
> kubectl -n rook-ceph get svc -l app=rook-ceph-mon -o yaml > rook-ceph-mon-services.yaml
> ```
> 
> 1. Edit the exported yaml files to make them consumable in a new cluster by removing cluster-specific properties. For example:
>    
>    * Remove `ownerReferences`, `resourceVersion`, `selfLink`, `uid`
>    * Add `namespace: rook-ceph` where missing
>    * For the services, remove the list properties and add a separator `---` between each of the services
> 2. Delete the Rook and/or K8s cluster (in my case I just deleted the Rook cluster.yaml, operator.yaml, and common.yaml to simulate a deleted cluster. It should have the same effect as bringing up a new k8s instance)
> 3. Create the new K8s cluster
> 4. Create the `rook-ceph` namespace and other basic resources: `kubectl create -f common.yaml`
> 5. Restore the Rook resources from the previous cluster
> 
> ```
> kubectl create -f rook-ceph-admin-keyring-secret.yaml
> kubectl create -f rook-ceph-mon-keyring-secret.yaml
> kubectl create -f rook-ceph-mon-services.yaml
> kubectl create -f rook-ceph-mon-endpoints-cm.yaml
> kubectl create -f rook-ceph-mon-secret.yaml
> ```
> 
> 1. Create the Rook cluster: operator.yaml and cluster.yaml
> 
> If all goes well, the mons will restore quorum, the existing devices will be discovered and corresponding OSDs started, and the cluster will start successfully. The only issue I saw at this point was that some of the mgr commands were timing out so it took longer than expected.
> 
> This should all be all capable of being automated. @Queuecumber want to give it a try after #4021 is merged?

I have followed these instructions. But in  my case no of OSD's are increasing for every restore. First I have started with one OSD. I have taken backup and when trying to restore it, no of OSD becomes 2. Please help me on this.

**Comment by grimmthetallest** [comment]
> It sounds to me like what we are talking about here is cluster adoption. Extending the ideas here to the most extreme case, all a user has is some number of OSD disks on some number of nodes from a prior Ceph cluster, and this cluster may not even have been a Rook cluster but a baremetal one. What is the minimum set of information needed to create a Ceph cluster that can use those disks as part of its storage? I think that would be the cluster FSID and the admin keyring. Or is it still required to have monitor data and the monitor keyring available? @leseb do you know?
> 
> ...Ceph cluster adoption is non-trivial, and that will be a longer burn project.

As someone with no prior knowledge of how Rook does things but presumes to have a decent understanding of Ceph internals, I'd like to point out that Ceph includes a scan tool for BlueStore devices that prints the cluster fsid and scanned OSD ID. There may be some additional prerequisites to scan the OSD storage devices if LVM or dmcrypt are also used.

[There's also some good documentation here](https://docs.ceph.com/docs/master/rados/troubleshooting/troubleshooting-mon/#recovery-using-osds) on how to completely rebuild all monitors offline directly from the OSDs.

Given the ability to set the cluster fsid and keyrings explicitly to match previous values, the next logical step is making sure the OSDs land on the right nodes and the correct crushmap is in place so that PGs can be found in the expected locations. Ideally you'd have an existing copy of the crushmap, either compiled or decompiled. However there seem to be [some methods available for forensically analyzing the Ceph RocksDB offline](https://files.fbausch.de/Bausch_Master-Thesis_Ceph-Forensics.pdf) to extract the OSD/CRUSH maps.

Essentially, if you are left with just a pile of OSDs and all the higher level configuration has gone up in smoke (this is still assuming the OSDs were still in the best case of fully synced and safely halted) you should in theory be able to take the following steps to restore.

1. Attach offline OSD device to temporary mount and gather needed data (cluster fsid, OSD/CRUSH maps)
2. Use OSD ids and CRUSH locations to map OSD paths with the correct host<>osd topology
3. Rebuild the monitor store from newly remapped OSD directories and ceph-objectstore-tool rebuild procedure
4. Bring cluster monitors online and achieve quorum
5. Run prime-osd-dir to prepare each OSD directory and activate all OSDs
6. Pray that nothing important (keyrings for cephx auth, ceph.conf customizations, osd or pool flags, all the additional MDS bits CephFS needs) ends up in a different state from the original cluster.

**Comment by travisn** [comment]
As a reference, we have a doc now for disaster recovery either when mon quorum is lost of the whole kubernetes cluster is lost. See the [disastery recovery guide](https://rook.io/docs/rook/v1.2/ceph-disaster-recovery.html). 

There is still more to simplify and document around this, so leaving this issue open...

**Comment by jeonggyuchoi** [comment]
@travisn
Hi travisn
I really thanks for your reply and contribution.

I have some problem in my rook-ceph cluster

I update K8s TLS cert key,
and I recevied message that rook-ceph tool
```
Error from server: error dialing backend: x509: cannot validate certificate for 172.XX.XX.48 because it doesn't contain any IP SANs
```
So, I try to update the key,
First delete secret operator, mgr, osd and restart failed osd , operator pod

But still have problem cannot connect the tool.

Accidentally, I delete the " rook-ceph-mon secret " // OMG -.-

How do i recover the rook-ceph-mon secret // .

I'm beginer / with rook-ceph v1.0.4 / k8s v1.14.5

**Comment by travisn** [comment]
@jeonggyuchoi What rook-ceph tool do you mean cannot connect? If it's the toolbox pod, you could delete the pod and see if it connects after restarted.

If you deleted the secret, do you still have any of the mon daemons running? The secret should be found inside the mon pods, in a config file perhaps somewhere under `/var/lib/ceph/mon`

**Comment by jeonggyuchoi** [comment]
@travisn  Thanks you for your comment ^^
1. Yes  I deleted the "toolbox" pod  and get a error message on pod "CreateContainerConfigError".
```
State:          Waiting
      Reason:       CreateContainerConfigError
    Ready:          False
    Restart Count:  0
    Environment:
      ROOK_ADMIN_SECRET:  <set to the key 'admin-secret' in secret '**rook-ceph-mon**'>  Optional: false
```

2. I check the  mon pods  too. /var/lib/ceph/mon 
```
-rw------- 1 root root   77 Aug 20  2019 keyring
-rw-r--r-- 1 root root    8 Aug 20  2019 kv_backend
-rw-r--r-- 1 root root    3 Aug 20  2019 min_mon_release
drwxr-xr-x 2 root root 4096 Feb 24 11:09 store.db
```

Actually I delete the secret rook-ceph-mon-token 
so 
```
NAME                           TYPE                                  DATA   AGE
default-token-tss22            kubernetes.io/service-account-token   3      188d
rook-ceph-admin-keyring        kubernetes.io/rook                    1      188d
rook-ceph-config               kubernetes.io/rook                    2      188d
rook-ceph-dashboard-password   kubernetes.io/rook                    1      188d
rook-ceph-mgr-a-keyring        kubernetes.io/rook                    1      188d
rook-ceph-mgr-token-25rm7      kubernetes.io/service-account-token   3      2d21h
rook-ceph-mons-keyring         kubernetes.io/rook                    1      188d
rook-ceph-osd-token-8v9rb      kubernetes.io/service-account-token   3      2d21h
rook-ceph-system-token-7gpqf   kubernetes.io/service-account-token   3      2d21h
```

I deleted like this secret.-.-;
**rook-ceph-mon**

Is that possible make it recovery.

**Comment by litaobowl** [comment]
> @travisn Would that prototype recover PVCs/PVs consuming Rook Ceph storage? The aim I think should be allowing using certain PVC (`mypersistentvolume`) from a Pod in the new Kubernetes cluster just as if it was deployed in the old cluster.

@travisn I have the same question as @albertdb, would the disaster recovery guide handle that?

**Comment by travisn** [comment]
> > @travisn Would that prototype recover PVCs/PVs consuming Rook Ceph storage? The aim I think should be allowing using certain PVC (`mypersistentvolume`) from a Pod in the new Kubernetes cluster just as if it was deployed in the old cluster.
> 
> @travisn I have the same question as @albertdb, would the disaster recovery guide handle that?

Today the disaster recovery guide does not document how to connect the PVs to the same application. As far as I'm aware, K8s does not have a way with dynamic provisioning to associate the same block image from a previous cluster with the new PV.

**Comment by jsmouret** [comment]
I was wondering if a PV identified by namespace/pv_name could be back by a directory namespace/pv_name on cephfs. In case of disaster, recreating the PV with the same namespace and name would end up using the same folder, hopefully kept by cephfs.

**Comment by travisn** [comment]
> I was wondering if a PVC identified by namespace/pvc_name could be backed by a directory namespace/pvc_name on cephfs. In case of disaster, recreating the PVC with the same namespace and name would end up using the same folder, hopefully kept intact by cephfs.

@jsmouret Yes, if you have a rwx volume with cephfs, you can mount it to another pod. Since you're using the same PVC, it would use the same folder as long as it is bound to the same PV. But whether it is rxo or rwx I believe you'll have the same issue with getting the PV re-created and bound to the expected PVC.

**Comment by jsmouret** [comment]
The idea was that a PV dynamically provisioned to satisfy a PVC `mynamespace/mypersistentvolume` should use a predictable directory path like `/mynamespace/mypersistentvolume` on cephfs instead of `/pv_generated_uuid`.
If the kubernetes cluster is lost, simply recreating a PVC with the same old name would provision a PV with the same old path and recover the data.

**Comment by travisn** [comment]
Closing this old issue since we have the disaster recovery guide already. More specific improvements are being tracked in other issues such as #3985 and #5781
